reliability-engineering/terraform/modules/concourse-monitoring/files/dashboards/concourse-slo.json

```diff
@@ -0,0 +1,417 @@
+              },
+              {
+                "color": "dark-green",
+                "value": 99.9
+              }
+            ]
+          },
+          "unit": "percentunit"
+        },
+        "overrides": []
+      },
+      "gridPos": {
+        "h": 8,
+        "w": 6,
+        "x": 0,
+        "y": 0
+      },
+      "id": 6,
+      "interval": "",
+      "options": {
+        "colorMode": "value",
+        "graphMode": "area",
+        "justifyMode": "auto",
+        "orientation": "auto",
+        "reduceOptions": {
+          "calcs": [
+            "mean"
+          ],
+          "fields": "/^errorRatio$/",
+          "values": false
+        }
+      },
+      "pluginVersion": "7.0.5",
+      "targets": [
+        {
+          "alias": "",
+          "apiMode": "Metrics",
+          "dimensions": {
+            "LoadBalancer": "app/cd-concourse-web/ba5029fcd9b9cde2"
+          },
+          "expression": "",
+          "hide": true,
+          "id": "status5xx",
+          "matchExact": true,
+          "metricName": "HTTPCode_ELB_5XX_Count",
+          "namespace": "AWS/ApplicationELB",
+          "period": "300",
+          "refId": "A",
+          "region": "eu-west-2",
+          "statistics": [
+            "Sum"
+          ]
+        },
+        {
+          "alias": "",
+          "dimensions": {
+            "LoadBalancer": "app/cd-concourse-web/ba5029fcd9b9cde2"
+          },
+          "expression": "",
+          "hide": true,
+          "id": "status4xx",
+          "matchExact": true,
+          "metricName": "HTTPCode_ELB_4XX_Count",
+          "namespace": "AWS/ApplicationELB",
+          "period": "300",
+          "refId": "B",
+          "region": "default",
+          "statistics": [
+            "Sum"
+          ]
+        },
+        {
+          "alias": "",
+          "dimensions": {
+            "LoadBalancer": "app/cd-concourse-web/ba5029fcd9b9cde2"
+          },
+          "expression": "",
+          "hide": true,
+          "id": "status2xx",
+          "matchExact": true,
+          "metricName": "HTTPCode_Target_2XX_Count",
+          "namespace": "AWS/ApplicationELB",
+          "period": "300",
+          "refId": "F",
+          "region": "default",
+          "statistics": [
+            "Sum"
+          ]
+        },
+        {
+          "alias": "",
+          "dimensions": {
+            "LoadBalancer": "app/cd-concourse-web/ba5029fcd9b9cde2"
+          },
+          "expression": "",
+          "hide": true,
+          "id": "status3xx",
+          "matchExact": true,
+          "metricName": "HTTPCode_ELB_3XX_Count",
+          "namespace": "AWS/ApplicationELB",
+          "period": "300",
+          "refId": "G",
+          "region": "default",
+          "statistics": [
+            "Sum"
+          ]
+        },
+        {
+          "alias": "",
+          "dimensions": {},
+          "expression": "1 - status5xx / (status2xx+status3xx+status4xx+status5xx)",
```

Comment:
I see this being something more like

```
(RequestCount-target5xx) / (RequestCount+elb5xx)
```

To explain:

- `RequestCount - target5xx` is the total number of successful (interesting) requests, because it's the total number of requests that a target responded to, that weren't a 5xx error.
- `RequestCount + elb5xx` is the total number of (interesting) requests, either because a target responded or because the ELB raised a 5xx
- `target5xx` and `elb5xx` need to be defined elsewhere in the obvious way
- I am defining interesting requests to be: requests that a target responded to, or requests that an ELB returned a 5xx for.  This excludes non-5xx ELB responses such as redirects (eg http -> https) or fixed responses (such as 404s for particular paths).  I think this is a feature not a bug.

Reply:
hmm I have mixed feelings here.

If we BUILT concourse, I would 100% agree with you. target 500s would indicate a problem with the application that we should fix. However we only RUN concourse, so the part of the service we provide here is largely scoped to ensuring the thing is up.

I really wanted to "availability" to capture the points where the load balancer was unable to serve a request. ie. a 502/503/504 from the load balancer almost certainly means that the request did not return from the web node. Which is very likely to be a problem with how we have configured something or a serious issue causing problems.

My concern with tracking 5xx from the target may provide us only with information that concourse itself is being crappy.

Maybe we could track both seperately and see how it pans out.... but I do believe tracking the non-target 5xx is a better metric for "we are ensuing that concourse is available" rather than "we are ensuring the concourse application itself is correct"

Reply:
Hmm I don't know enough about concourse itself to know if 5xxs from concourse are likely to be logic error (not much we can do) or resource issue (we should care).  For example, can concourse return 502 or 504 if ATC is up but TSA/garden/reaper/beacon is down somewhere?

Reply:
taking this on board and after the continued weirdness with requestCount and following the other comments about ignoring 3xx and 4xx the expression has been simplified to essentially:

```
1 - (elb5xx + target5xx) / ( elb5xx + target5xx + target2xx )
```

doesn't change the graph much, but as you say should catch the target5xx too